pkg/math/arith/modulus.go

```diff
@@ -0,0 +1,77 @@
+package arith
+
+import (
+	"github.com/cronokirby/safenum"
+)
+
+// Modulus wraps a safenum.Modulus and enables faster modular exponentiation when
+// the factorization is known.
+// When n = p⋅q, xᵉ (mod n) can be computed with only two exponentiations
+// with p and q respectively.
+type Modulus struct {
+	// represents modulus n
+	*safenum.Modulus
+	// n = p⋅p
+	p, q *safenum.Modulus
+	// pInv = p⁻¹ (mod q)
+	pNat, pInv *safenum.Nat
+}
+
+// ModulusFromN creates a simple wrapper around a given modulus n.
+// The modulus is not copied.
```

Comment:
To note that not copying is perfectly fine, since the methods on Modulus are all immutable!